Rewrite prof dumping

yinan1048576 opened the pull request:
The high level goals of the rewrite are two folds:
- Separate file operations out from the core profiling logic.
  - On a logical level, the module involving file operations should be considered as an auxiliary library to jemalloc, not as part of the core jemalloc. File operations are generally platform dependent. Core jemalloc should interact with the file system indirectly, via write back handles. The current `malloc_stats_print()` sets a good example of this design principle, and it empirically led to improved clarity and simplicity, as compared to the prof dumping setup.
- Test file operation failures thoroughly.
  - On a logical level, the tests are not for the core jemalloc, but for the file handling auxiliary module. A test set with a high / full test coverage is not trivial to write - it even feels reasonable to me to push the file handling part completely out of jemalloc for the benefit of not having to deal with the tests. However, in the short- to mid- term, keeping the file handling part within jemalloc would be easier for jemalloc-dependent applications, as compared to having a separate library, and we should take the responsibility of testing.

I'm planning to achieve the goals in multiple stages. At this moment, I think the highest priority work item is to first get the tests. On the one hand, it looks to me like we don't have any tests for file operation failures. On the other hand, the tests enable me to clearly state the functionality objectives, and my subsequent development work can first focus on making the tests pass, i.e. the functionality fixing part, and then on keeping the tests pass, i.e. the refactoring part.

In this first batch of commits I'm submitting, I'm doing:
- the minimal structural refactoring for the benefit of being able to write tests;
- writing tests that can cover (I hope) *all* possibilities of file operation failures.

Specifically, the functionality objectives I'd like the dumping logic to achieve are:
- If we fail to open the dump file, then we don't write anything and return failure.
  - This is trivial and is already achieved by the current logic. We just need a test to be able to guard it.
- If we fail to open the mapped libraries file, then we don't dump the mapped libraries section and return success.
  - This is *not* the case in the current logic: we return failure when the maps file cannot be opened. I think this is wrong, but please correct me otherwise. An extreme example is Windows: we don't even try to locate the maps file, which means prof dumping always fails on Windows according to our current logic. More generally, I think the mapped libraries are a somewhat supplementary part of the dump, and can be treated as a "best effort" rather than a contract.
- If one file writing operation fails, then we don't try to write any more (either the current content or future content) and return failure.
  - This is *not* the case in the current logic: we may still write future content after a failure, though we don't try to write the current content again. I think this is more clearly wrong. File writing failures might be transient, so we may end up messing up the dump file content. The decision of not trying to write the same content again is reasonable; otherwise the logic would be more complicated yet not being able to make any better guarantee.

Before submitting any fix, I want to first test the tests, i.e. make sure that my new tests fail on all platforms as long as prof is turned on, and also that the tests fail in the ways I expect; otherwise I'll need to fix the tests first. After that, I'll submit main code changes in batches to first fix the functionality and then refactor the code structure.

yinan1048576 commented:
Failures are observed as expected. Interesting to find that we don't have prof turned on for any CI runs on Windows. Is this intended?

davidtgoldblatt commented:
> Is this intended?

Yes -- we don't have any backtracking libraries supported in windows.

yinan1048576 commented:
Submitting the fix commit and hopefully all tests pass. Afterwards I'll squash this commit with the unit test commit.

The fix has a behavior change in case a file opening / writing failure is encountered:
- Previously, the intended behavior is to immediately stop and return a failure. (I call it intended behavior because it's not strictly followed in reality, but the intention is clear.)
- Now, the write callback function notes the failure down, and it will not try to write when further being called; the core dumping logic is made to be unaware of the failure and will keep calling the callback function.

The cost is that the core dumping logic would keep traversing through all profiling content, which makes it slower than ideal; the benefit, as a direct consequence, is that the traversing logic can be greatly simplified, as is demonstrated by the new commit. I think the benefit outweighs the cost because:
- Failures are not that common, and when they happen, "failing performance" is usually not the top priority.
- It's not trivial to strictly stick to the "fail-and-immediately-stop" policy. The code would be much harder to read, and we might still miss some edge cases even if we try hard (like what we did).
- `malloc_stats_print()` follows the "keep-traversing" policy.

davidtgoldblatt commented on src/prof_data.c:
```diff
@@ -56,6 +56,20 @@ static ckh_t bt2gctx;
  */
 static prof_tdata_tree_t tdatas;
 
+/* The following are needed for dumping and are protected by prof_dump_mtx. */
```
The intended end state for this code is for this code is to rewrite it in terms of (e.g.) the buffered writer API, right?

Rather than passing these as globals, why not pull them into a `prof_output_ctx_t` or something that contains them as well as `prof_dump_fd`, and we can just propagate a pointer to that context struct everywhere. On its own, this change is sort of a sideways move; it highlights the non-error case better, but at the cost of making the flow of data much harder to reason about.

yinan1048576 replied:
> The intended end state for this code is for this code is to rewrite it in terms of (e.g.) the buffered writer API, right?

Yes.

> Rather than passing these as globals, why not pull them into a `prof_output_ctx_t` or something that contains them as well as `prof_dump_fd`, and we can just propagate a pointer to that context struct everywhere. On its own, this change is sort of a sideways move; it highlights the non-error case better, but at the cost of making the flow of data much harder to reason about.

Good point. I'll have that in the restructuring commits. For now the commit is more for the functionality change, and I'll put up commits for restructuring later (which will keep the functionality exactly the same and all the tests passing), but thanks for the early feedback.

yinan1048576 commented:
Rebased, squashed the fix commit with the unit test commit, and added a few more commits for restructuring the code. The last one creates a `prof_shell` module and I'm first moving the system thread name reading to it. The new module will be the place all platform dependent code in profiling goes to, which will include all the file and file name operations. I'm pausing here to get some early feedback on the `prof_shell` naming.

Haven't yet addressed the review feedback. Will do so in the next batch of commits.

yinan1048576 commented:
Minor correction.

davidtgoldblatt commented:
> `The new module will be the place all platform dependent code in profiling goes to, which will include all the file and file name operations. I'm pausing here to get some early feedback on the prof_shell naming.`

It's not obvious to me what "shell" here refers to (e.g. bash shell? I think in general we expect the thread name to get set in other ways). Maybe `prof_sys`?

yinan1048576 commented:
`prof_sys` sounds better. Updated.

yinan1048576 commented:
Rebase.

yinan1048576 commented:
Rebase, and added one final commit to migrate to buffered writer. Now the PR is ready for review.

> Rather than passing these as globals, why not pull them into a `prof_output_ctx_t` or something that contains them as well as `prof_dump_fd`, and we can just propagate a pointer to that context struct everywhere. On its own, this change is sort of a sideways move; it highlights the non-error case better, but at the cost of making the flow of data much harder to reason about.

I thought about it again and decided not to do it. The flow of data is hard in the sense that there's a loop "`prof_sys` -> `prof_data` -> `buffered_writer` -> `prof_sys`", but any place involving the usage of the `buffered_writer` module has a loop by design. Also, the loop is orthogonal to whether we define a struct for the dumping-relevant variables or not - they're all private to `prof_sys`.

davidtgoldblatt commented:
> I thought about it again and decided not to do it. The flow of data is hard in the sense that there's a loop "prof_sys -> prof_data -> buffered_writer -> prof_sys", but any place involving the usage of the buffered_writer module has a loop by design. Also, the loop is orthogonal to whether we define a struct for the dumping-relevant variables or not - they're all private to prof_sys.

It's not really about a circular dependency (which I agree is fine here; taking an argument as a callback isn't really quite the same as just forced calls). it's about being able to trace how information gets from callers to callees. The caller here wants to pass information (the file descriptor, etc.) to callees. We should do that through function arguments rather than as globals without a strong reason not to; having the behavior of a function change outside of its arguments can cause hard-to-reason-about interactions.

yinan1048576 commented:
Rebase, and replace all global variables with explicit parameter passing. Also made some minor simplifications.

yinan1048576 commented:
Slight improvement.

yinan1048576 commented:
Minor revision based on #1823.

yinan1048576 commented:
Rebase on top of #1823.

yinan1048576 commented:
Rebase.

yinan1048576 commented:
Rebase.

yinan1048576 commented:
Rebase.

yinan1048576 commented on src/prof_data.c:
```diff
@@ -492,17 +492,24 @@ prof_dump_open(bool propagate_err, const char *filename) {
 }
 
+#ifdef JEMALLOC_JET
+prof_dump_write_file_t *JET_MUTABLE prof_dump_write_file = malloc_write_fd;
```
Do you mean that these file pointers should always be exposed, regardless of whether the environment is `JEMALLOC_JET` or not? For that, the corresponding declarations in the headers files will also be unconditionally exposed.

davidtgoldblatt replied:
Yeah, just exposing everything is fine (everything still gets name-mangled, so it's not like, exposed to the user).

In the long-run, it would help us on a few axes to be able to kill off the test/non-test variants entirely. That probably won't be totally possible without C++/some templatizing, but in the meantime we at least don't want to make things harder in that dimension.

yinan1048576 commented on src/prof_data.c:
```diff
@@ -1,4 +1,4 @@
-#define JEMALLOC_PROF_C_
+#define JEMALLOC_PROF_DATA_C_
```
Good to know. Is there any platform / environment where these are still useful? Otherwise we should delete them all.

davidtgoldblatt replied:
I *believe* we've killed off all the places that these changed. I'm on board with deleting them.

yinan1048576 commented:
Revise according to feedback (by rebase).

yinan1048576 commented:
One more minor change, in addition to the requested revisions: define a smaller buffer size for testing directly in `prof_types.h` to reduce noise.

yinan1048576 commented:
Rebase. Hopefully the AppVeyor installation issue is gone now.

Also added two more commits, which get rid of some unnecessary code and also facilitate testing for #1777.

Ready for final round of review.

yinan1048576 commented:
Revised so that the smaller buffer size is conditioned on debug build rather than on test build, so as to get rid of any difference between test and non-test builds.
